src/WinstonGelfTransporter.ts

```diff
@@ -0,0 +1,137 @@
+import TransportStream from 'winston-transport';
+import log from 'gelf-pro';
+import os from 'os';
+
+type TransporterOptions = {
+  // Winston transporter options
+  level?: string;
+  silent?: boolean;
+  handleExceptions?: boolean;
+  // Graylog communication options
+  version?: string;
+  host?: string;
+  port?: number;
+  protocol?: string;
+  hostName?: string;
+  additional?: Object;
+};
+
+const MIN_PORT: number = 0;
+const MAX_PORT: number = 65535;
+
+export default class WinstonGelfTransporter extends TransportStream {
+  private options: TransporterOptions;
+
+  private logLevels = Object({
+    emerg: 0,
+    alert: 1,
+    crit: 2,
+    error: 3,
+    warn: 4,
+    notice: 5,
+    info: 6,
+    debug: 7
+  });
+
+  private objectSerializer: (o: object) => string;
+
+  readonly gelfClient: any;
+
+  constructor(options: TransporterOptions) {
+    super(options);
+    this.options = options;
+    const logConfig = Object({ fields: {} });
+
+    // Update protocol if provided
+    if (options.protocol) {
+      logConfig.adapterName = options.protocol;
+    }
+    // Update host and port information if provided
+    if (options.port || options.host) {
+      logConfig.adapterOptions = {};
+      if (
+        options.port &&
+        options.port >= MIN_PORT &&
+        options.port < MAX_PORT
+      ) {
+        logConfig.adapterOptions.port = options.port;
+      }
+      if (options.host) {
+        logConfig.adapterOptions.host = options.host;
+      }
+    }
+    logConfig.fields = options.hostName
+      ? { host: options.hostName }
+      : { host: os.hostname() };
+    if (options.additional) {
+      logConfig.fields = {
+        ...logConfig.fields,
+        ...options.additional
+      };
+    }
+    this.objectSerializer = JSON.stringify;
+
+    log.setConfig(logConfig);
+    this.gelfClient = log;
+  }
+
+  /**
+   * This function will return a network connector
+   * for sending GELF messages.
+   * @returns A TCP Client for graylog.
+   */
+  getNetworkConnector(): any {
+    return this.gelfClient;
+  }
+
+  getLogLevel(level: string): number {
+    return this.logLevels[level];
+  }
+
+  setObjectSerializer(serializer: (o: object) => string): void {
+    this.objectSerializer = serializer;
+  }
+
+  getLogMessage(info: any): string | Error {
```

Comment:
Well, info should never be of type string. According to Winston it should send an object.
The additional objects contained in info could be named anything, and you can't assume a member of the object.
For example:
in the case of the log statement
`log.info('user logged in with certain credentials', userObj)`
where `userObj` looks like
`{name: 'Me', email: 'anEmail@example.com'}`
the winston logger will send the `log` method (and therefore this method)
`{message: 'user logged in with certain credentials', name: 'Me', email: 'anEmail@example.com'}`
Let's say we set the the type def for info to object.
If I were to then say `const { message } = info` typescript will not compile. I would be assuming that there is member called message.
I could do something along the lines of
```
getLogMessage(info: {
    message: string | object | Error;
    [keys: string]: any;
  }
)
```
but since the `TransportStream` library itself expresses the functional definition as `log(info: any)`, I'd be then passing an `any` typed object to our own custom function parameter type.

Reply:
OK, makes sense.